pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,622 @@
+		ownerNode, ok := p.uidToNode[owner.UID]
+		if !ok {
+			// Create a "virtual" node in the graph for the owner if it doesn't
+			// exist in the graph yet. Then enqueue the virtual node into the
+			// dirtyQueue. The garbage processor will enqueue a virtual delete
+			// event to delete it from the graph if API server confirms this
+			// owner doesn't exist.
+			ownerNode = &node{
+				identity: objectReference{
+					OwnerReference: owner,
+					Namespace:      n.identity.Namespace,
+				},
+				dependents: make(map[*node]struct{}),
+			}
+			p.uidToNode[ownerNode.identity.UID] = ownerNode
+			p.gc.dirtyQueue.Add(ownerNode)
+		}
+		ownerNode.dependents[n] = struct{}{}
+	}
+}
+
+// insertNode insert the node to p.uidToNode; then it finds all owners as listed
+// in n.owners, and adds the node to their dependents list.
+func (p *Propagator) insertNode(n *node) {
+	p.uidToNode[n.identity.UID] = n
+	p.addDependentToOwners(n, n.owners)
+}
+
+// removeDependentFromOwners remove n from owners' dependents list.
+func (p *Propagator) removeDependentFromOwners(n *node, owners []metatypes.OwnerReference) {
+	for _, owner := range owners {
+		ownerNode, ok := p.uidToNode[owner.UID]
+		if !ok {
+			continue
+		}
+		delete(ownerNode.dependents, n)
+	}
+}
+
+// removeNode removes the node from p.uidToNode, then finds all
+// owners as listed in n.owners, and removes n from their dependents list.
+func (p *Propagator) removeNode(n *node) {
+	delete(p.uidToNode, n.identity.UID)
+	p.removeDependentFromOwners(n, n.owners)
+}
+
+// TODO: profile this function to see if a naive N^2 algorithm performs better
+// when the number of references is small.
+func referencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerReference) (added []metatypes.OwnerReference, removed []metatypes.OwnerReference) {
+	oldUIDToRef := make(map[string]metatypes.OwnerReference)
+	for i := 0; i < len(old); i++ {
+		oldUIDToRef[string(old[i].UID)] = old[i]
+	}
+	oldUIDSet := sets.StringKeySet(oldUIDToRef)
+	newUIDToRef := make(map[string]metatypes.OwnerReference)
+	for i := 0; i < len(new); i++ {
+		newUIDToRef[string(new[i].UID)] = new[i]
+	}
+	newUIDSet := sets.StringKeySet(newUIDToRef)
+
+	addedUID := newUIDSet.Difference(oldUIDSet)
+	removedUID := oldUIDSet.Difference(newUIDSet)
+
+	for uid := range addedUID {
+		added = append(added, newUIDToRef[uid])
+	}
+	for uid := range removedUID {
+		removed = append(removed, oldUIDToRef[uid])
+	}
+	return added, removed
+}
+
+func shouldFinalize(e event, accessor meta.Object) bool {
+	// The delta_fifo may combine the creation and update of the object into one
+	// event, so we need to check AddEvent as well.
+	if accessor.GetName() == "test.rc.1" {
+		fmt.Printf("CHAO: test.rc.1, e.obj=%#v\n", e.obj)
+		fmt.Println("CHAO: test.rc.1, accessor.GetDeletionTimestamp()=", accessor.GetDeletionTimestamp())
+		fmt.Println("CHAO: e.oldObj=e.oldObj")
+	}
+	if e.oldObj == nil {
+		if accessor.GetDeletionTimestamp() == nil {
```

Comment:
This line handles the case where the DeletionTimestamp is set at the first time GC sees it